telemetry/telemetry.proto

```diff
@@ -4,6 +4,72 @@ import 'nanopb.proto';
 
 option java_outer_classname="TelemetryProto";
 
-message Telemetry {
-  optional int32 timestamp_ms = 1;
+//A DeviceInfoMessage should only be sent if a request is received from the telemetry viewer (ex. if the
+//		viewer sees a device it hasn't registered yet)
+//Otherwise, the viewer will just reference its config files (stored offline) for this information
+//Note: formerly NameIDMessage
+message DeviceInfoMessage { //send the device ID with its corresponding information
+	enum READING_TYPE {
+		VOLTAGE = 1;
+		CURRENT = 2;
+		TEMPERATURE = 3;
+		MCC = 4;
+		DRIVER_CONTROL = 5; //i.e. pedal board
+		OTHER = 127;
+	}
+	optional uint32 id = 1;
+	optional READING_TYPE readingType = 2 [default = OTHER];
+	optional string name = 3 [(nanopb).max_size = 32];
+	optional uint32 expectedSampleDelay = 4; //expected time between sample readings
+	optional uint32 longestAllowedSampleDelay = 5;
+	optional uint32 numAggregateSamples = 6; //= numSamples in StatisticalAggregate (below)
+	optional string units = 7 [(nanopb).max_size = 4]; //max of four letters (mV, A, degC, mph, etc.)
+	optional uint32 minSensorLimit = 8;
+	optional uint32 maxSensorLimit = 9;
+	optional uint32 resolution = 10;
+	optional uint32 accuracy = 11;
+}
+
+message StatisticalAggregate {
+	optional uint32 numSamples = 1;
+	optional uint32 min = 2;
+	optional uint32 max = 3;
+	optional uint32 mean = 4;
+	optional uint32 stdev = 5;
+}
+
+message MeasurementAggregate {
+	optional uint32 id = 1; //id of the device being read
+	optional StatisticalAggregate reading = 2;
+	optional StatisticalAggregate sampleDelay = 3; //statistics on the times between samples
+												   //(largely for debugging, heartbeat/reading failures
+												   //will be send as InfoMessages (below))
+}
+
+message InfoMessage {
+	enum MESSAGE_TYPE { //TODO: expand the list of error types for quick identification
+		INFO_MESSAGE = 1; //right now, this is the only non-error message type
```

Comment:
In Datalogger, I use InfoMessage string purely as a fallback, for non-structured or very-infrequent-not-warranting-a-dedicated-message data. I don't think it makes sense to mix that with the more structured messages you have here (for example, ERROR_WARNING and ERROR_PASSIVE are CAN-specific messages, and HEARTBEAT_FAILURE, READING_FAILURE, and DATA_OVERRUN are structured.

If you wanted to associate a severity with general string messages (which can be a good idea, especially if a person is parsing in realtime), the traditional levels are "debug", "info", "warning", and "error" - with the last one usually before the program crashes.